File: modules/react-native-mlkit-text-recognition/ios/RNMLKitTextRecognitionModule.swift
import ExpoModulesCore
import MLKitCommon
import MLKitTextRecognition
import RNMLKitCore

public class RNMLKitTextRecognitionModule: Module {
    let ERROR_DOMAIN: String = "red.infinite.RNMLKit.RNMLKitTextRecognitionModuleErrorDomain"
    
  public func definition() -> ModuleDefinition {
    Name("RNMLKitTextRecognition")
    
    AsyncFunction("recognizeText") { (imagePath: String, promise: Promise) in
      let logger = Logger(logHandlers: [createOSLogHandler(category: Logger.EXPO_LOG_CATEGORY)])
      logger.info("RNMLKit", "recognize text: Recognizing text from image: \(imagePath) ")
      
      let options = TextRecognizerOptions()
      let textRecognizer = TextRecognizer.textRecognizer(options:options)
      
      let image = try RNMLKitImage(imagePath: imagePath)
      
      Task {
        do {
          let result = try await textRecognizer.process(image.visionImage)
          
          promise.resolve(mapTextToRecord(result))
        } catch {
          promise.reject(
              NSError(domain: ERROR_DOMAIN, code: 1, userInfo: [NSLocalizedDescriptionKey: "Error occurred: \(error)"])
          )
        }
      }
    }
  }
}

File: modules/react-native-mlkit-text-recognition/android/src/main/java/red/infinite/reactnativemlkit/textrecognition/RNMLKitTextRecognitionModule.kt
package red.infinite.reactnativemlkit.textrecognition

import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.common.InputImage
import kotlinx.coroutines.runBlocking
import red.infinite.reactnativemlkit.core.RNMLKitImage
import android.net.Uri

class RNMLKitTextRecognitionModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("RNMLKitTextRecognition")

    AsyncFunction("recognizeText") { imagePath: String, promise: Promise ->
      runBlocking {
        val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)

        val image: InputImage = RNMLKitImage(Uri.parse(imagePath), appContext.reactContext!!).image

        recognizer.process(image)
          .addOnSuccessListener { visionText ->
            promise.resolve(mapTextToRecord(visionText))
          }
          .addOnFailureListener { e ->
            promise.reject(
              CodedException(
                "RNMLKitTextRecognitionModule - Error: ${e.message}", e
              )
            )
          }
      }
    }
  }
}
